### PR TITLE
expando thing on jsValue for jsArray

### PR DESCRIPTION
**Pull request by dam0vm3nt**
As suggested by @jakemac53 this will fix auto notification support for jsArray.

### CODE REVIEW

**Comment by dam0vm3nt** [comment]
I signed it!

File: lib/src/convert.dart [hunk 1]
```diff
@@ -21,13 +21,19 @@ abstract class JsProxyInterface {
   JsObject get jsProxy;
 }
 
+Expando<JsArray> jsArrayExpando = new Expando<JsArray>();
```
Comment (jakemac53) [inline lib/src/convert.dart]: lets make this private

File: lib/src/convert.dart [hunk 2]
```diff
@@ -21,13 +21,19 @@ abstract class JsProxyInterface {
...
+
 /// Converts a dart value to a js value, using proxies when possible.
 /// TODO(jakemac): Use expando to cache js arrays that mirror dart lists?
 dynamic jsValue(dartValue) {
   if (dartValue is JsProxyInterface) {
     return dartValue.jsProxy;
   } else if (dartValue is Iterable) {
-    var newList = new JsArray.from(dartValue.map((item) => jsValue(item)));
+    var newList = jsArrayExpando[dartValue];
+    if (newList==null) {
```
Comment (jakemac53) [inline lib/src/convert.dart]: dartformat please :)

**Comment by jakemac53** [comment]
@sigmundch What do you think about this? The primary risk here is if the two lists could get out of sync there is no way to recover from it, since the JsArray is now permanently tied to the Dart List.

**Comment by sigmundch** [comment]
mmm interesting - it's a valid concern. Wouldn't you run into other issues if you get out of sync, even without this?

I also have a couple questions:
* do you think we should do the same thing for Maps?
* should we also do this on the way back when we convert to Dart (set the expando in `dartValue` when we first create a Dart list for an existing JsArray)?

File: lib/src/convert.dart [hunk 3]
```diff
@@ -21,13 +21,19 @@ abstract class JsProxyInterface {
...
+      newList = new JsArray.from(dartValue.map((item) => jsValue(item)));
+      jsArrayExpando[dartValue] = newList;
+    }
     addDartInstance(newList, dartValue);
```
Comment (dam0vm3nt) [inline lib/src/convert.dart]: may be `addDartInstance` should be pulled inside `if` statement.

**Comment by jakemac53** [comment]
> do you think we should do the same thing for Maps?

Yes, otherwise it would be inconsistent. @dam0vm3nt  can you add that into this PR?

> should we also do this on the way back when we convert to Dart (set the expando in dartValue when we first create a Dart list for an existing JsArray)?

A bit riskier, but I still think we should. @dam0vm3nt can you do this as well?

**Comment by dam0vm3nt** [comment]
done. check it out.

**Comment by jakemac53** [comment]
Nice, can you also resolve the CLA issue? I think the problem is that your git commits are using a different email than your github account.

**Comment by dam0vm3nt** [comment]
> Nice, can you also resolve the CLA issue? I think the problem is that your git commits are using a different email than your github account.

I've added the email address to my github account, don't know if this is sufficient...

**Comment by dam0vm3nt** [comment]
Ok, googlebot just answered :)

**Pull request merged**